pkg/servicemesh/federation/server/routing_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"testing"
+
+	"istio.io/istio/pkg/cluster"
+	"istio.io/istio/pkg/config/labels"
+	federationmodel "istio.io/istio/pkg/servicemesh/federation/model"
+)
+
+func TestCreateResourceName(t *testing.T) {
+	testCases := []struct {
+		name                 string
+		mesh                 cluster.ID
+		source               federationmodel.ServiceKey
+		expectedResourceName string
+	}{
+		{
+			name: "standard mesh name + standard namespace + standard service name",
+			mesh: "mesh",
+			source: federationmodel.ServiceKey{
+				Namespace: "bookinfo",
+				Name:      "productpage",
+			},
+			expectedResourceName: "fed-exp-productpage-bookinfo-mesh",
+		},
+		{
+			name: "long mesh name",
+			mesh: "myveryveryverylongmesh",
+			source: federationmodel.ServiceKey{
+				Namespace: "bookinfo",
+				Name:      "productpage",
+			},
+			expectedResourceName: "fed-exp-productpage-bookinfo-myveryveryverylongmesh",
+		},
+		{
+			name: "long namespace",
+			mesh: "mesh",
+			source: federationmodel.ServiceKey{
+				Namespace: "bigbigbigbigbookinfo",
+				Name:      "productpage",
+			},
+			expectedResourceName: "fed-exp-productpage-bigbigbigbigbookinfo-mesh",
+		},
+		{
+			name: "long service name",
+			mesh: "mesh",
+			source: federationmodel.ServiceKey{
+				Namespace: "bookinfo",
+				Name:      "productpage-rest-private",
+			},
+			expectedResourceName: "fed-exp-productpage-rest-private-bookinfo-mesh",
+		},
+		{
+			name: "long mesh name + long namespace + long service name => truncated + hashed",
+			mesh: "test-remote",
+			source: federationmodel.ServiceKey{
+				Namespace: "bookinfo-production",
+				Name:      "productpage-rest-private",
+			},
+			expectedResourceName: "fed-exp-productpage-rest-private-bookinfo-production-teefc2b162",
```

Comment:
Could we call a hash function in this test case instead of hard coding the hash value "teefc2b162" ?

Reply:
Wouldn't that just make the test less readable for no benefit? I mean, you'd have to concatenate the full name in the exact same way as the code does it.

Reply:
well, if we assume this value is fixed and always generated by using hash length 8, then that's fine here. 
If we need to update a hash length value or the implementation of the hashResourceName in the future, this test case is not flexible to test a hash length 9 case.

Reply:
I've just thought of an alternative approach. What if `expectedResourceName` was a RegExp. In the example above, the regexp would be: `"fed-exp-productpage-rest-private-bookinfo-production-te[a-f0-9]{8}"` indicating that we only care about the part that was truncated and that the part at the end is an 8-character hex value, but we don't care about the actual value.

But then again, this would just make the test more complicated. Someone seeing the test for the first time would have a harder time understanding the test compared to what the test looks like now. That's why I think the current form is perfectly fine.